src/tracing/node_trace_writer.cc

```diff
@@ -138,8 +138,10 @@ void NodeTraceWriter::FlushPrivate() {
 
 void NodeTraceWriter::Flush(bool blocking) {
   Mutex::ScopedLock scoped_lock(request_mutex_);
-  if (!json_trace_writer_) {
-    return;
+  {
+    Mutex::ScopedLock stream_mutex_lock(stream_mutex_);
```

Comment:
Is it necessary to nest locks?<sup>1</sup> And if so, can you perhaps add a comment explaining why? It's currently kind of hard to eyeball whether this is correct or a deadlock waiting to happen.

It would be nice to have an abstraction that enforces that locks are only taken out in a specific order so you don't run into AB-BA deadlocks.

<sup>1</sup> I think the answer is 'yes' because otherwise there's a race window.

Reply:
> Is it necessary to nest locks?<sup>1</sup>

I think so, yes, for the reason you mentioned. But I am not a hundred percent sure either. I have a hard time figuring out this code myself as well.

> And if so, can you perhaps add a comment explaining why? It's currently kind of hard to eyeball whether this is correct or a deadlock waiting to happen.

Added a comment; I did check that the locks are never required in another order, so this shouldn’t be making anything *worse*…